README.md
# pwget

pwget is a *stateless password manager*, a tool that generates all your
passwords from a single master password. Its main difference from similar tools
is that it maintains a list of revoked passwords. You can revoke a password if
a service that you use has had their password database leaked, or if you just
like to change your passwords regularly.

Of course, one can argue that the revocation list makes pwget
not-stateless-anymore, but I still consider it mostly stateless because the
revocation list can always be recovered by starting from an empty list,
generating passwords, and revoking those that don't work until you get to the
first non-revoked password.

## Security considerations

**WARNING: I'm not a professional cryptographer.** I deny all responsibility if
you use this code and then lose your passwords or identity.

Having said that, I consider the main attack vector to be identity theft
through sniffing the master password, e.g. via a keylogger. That attack vector
exists for all password managers that use a master password.

But for a stateless password manager, a master password cannot easily be
changed, unless you are willing to change all passwords derived from it. I
don't see a practical difference, though: If your master password is
compromised, you should always assume all derived passwords compromised as
well, since the adversary has likely already obtained a copy of your password
store (if your password manager maintains one).

In the end, the same advice as always holds true: Know your threat model, and
carefully weigh security against utility.

## Installation

```bash
# option 1
make
sudo make install

# option 2
make
ln -s $PWD/build/pwget ~/bin # or wherever you put your tools
```

## Usage

To generate a password,

```bash
pwget $DOMAIN
```

where `$DOMAIN` identifies the service that you're trying to log on. It's your
job to come up with a consistent scheme for these `$DOMAIN` values. pwget does
not enforce anything. The following three, for example, will generate
completely different passwords:

```bash
pwget example.com
pwget www.example.com
pwget johndoe@example.com
```

To revoke a password,

```bash
pwget --revoke $DOMAIN
```

Then generate the next password with `pwget $DOMAIN`. The list of revoked
passwords is stored at `$HOME/.pwget-revocation`, which contains the SHA256
hash of every revoked password. pwget will always show a list of all revoked
passwords that it encounters, so you can recover from an undesired revocation
easily.

### Maximum length restriction

Some stupid services enforce a maximum length of passwords. You should stop
using these services, but if you absolutely cannot, you can pass the maximum
password length as a second argument to truncate the generated password:

```bash
# stupidbank.com rejects passwords longer than 20 chars
pwget stupidbank.com 20
```

## Algorithm

In pseudo-code:

```
func getPassword(masterPassword, domain) {
    for iteration in 0..infinity {
        salt = iteration + ":" + domain
        pw = scrypt(password, salt, N = 2^16, r = 8, p = 16, keylength = 32 bytes)
        if pw is not revoked {
            return z85_encode(pw)
        }
    }
}
```

Where `scrypt()` is the SCrypt key derivation function, and `z85_encode()`
encodes the bytearray obtained from `scrypt()` to text using the [Z85
encoding](http://rfc.zeromq.org/spec:32/Z85), a variant of Base85.

Encoding passwords in Z85 ensures that the overwhelming majority of passwords
include lowercase letters, uppercase letters, numbers and symbols, like a lot
of stupid websites require.


CHANGELOG.md
# v2.1.0 (2022-05-18)

Changes:

- When stdin is not a tty, do not show a prompt; just read the master password
  from stdin directly. This mode of operation is useful when calling pwget from
  a script or other automated process.
- Add go.mod file to work with newer Go versions.
- Update all dependencies.

# v2.0 (2017-12-26)

**Backwards-incompatible changes:**

- Passwords are encoded in Base-85 (Z85) instead of Base-16 (hexadecimal).
  This increases entropy density, and ensures that more diverse character sets
  (i.e. uppercase letters and symbols) are used, as required by some websites.
  However, this means that entirely different password strings are generated.

Changes:

- Rename binary from `pwget` to `pwget2` to allow parallel installation with
  1.x version (for transitioning passwords one after one).
- Windows is now a supported target platform.
- Add optional length argument that truncates passwords to comply with maximum
  password length requirements of stupid services.

# v1.2 (2017-06-08)

Bugfixes:

- Fix display of prompt when stdout is piped.

# v1.1 (2017-06-07)

Changes:

- Remove openssl-1.0 dependency.

# v1.0 (2016-03-20)

Initial release.


main.go
/*******************************************************************************
*
* Copyright 2016 Stefan Majewsky <majewsky@gmx.net>
*
* This program is free software: you can redistribute it and/or modify it under
* the terms of the GNU General Public License as published by the Free Software
* Foundation, either version 3 of the License, or (at your option) any later
* version.
*
* This program is distributed in the hope that it will be useful, but WITHOUT
* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
* FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
* details.
*
* You should have received a copy of the GNU General Public License along with
* this program. If not, see <http://www.gnu.org/licenses/>.
*
*******************************************************************************/

package main

import (
	"crypto/sha256"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
	"syscall"

	"github.com/tilinna/z85"
	"golang.org/x/crypto/scrypt"
	"golang.org/x/crypto/ssh/terminal"
)

func main() {
	//check arguments
	domain, length, doRevoke := ParseArguments()

	//load revocation list
	isSHA256OfRevokedPassword, err := LoadRevocationList()
	FailOnError("load revocation list", err)

	//get master password
	masterPassword, err := GetMasterPassword()
	FailOnError("password query", err)

	//now we're all set, derive the password
	var passwordStr, hashStr string
	for iteration := 0; true; iteration++ {
		//get the password for this iteration
		salt := []byte(strconv.Itoa(iteration) + ":" + domain)
		password := Scrypt(masterPassword, salt)
		passwordStr = z85EncodeToString(password)

		//this is the correct password, unless it has been revoked
		hash := sha256.Sum256([]byte(passwordStr))
		hashStr = z85EncodeToString(hash[:])
		if isSHA256OfRevokedPassword[hashStr] {
			os.Stderr.Write([]byte("\x1B[37mhash:" + hashStr + " is revoked\x1B[0m\n"))
			continue
		}
		break
	}

	//now, we either print or revoke the thing
	if doRevoke {
		os.Stderr.Write([]byte("Revoking hash:" + hashStr + "\n"))
		err := AppendToRevocationList(hashStr)
		FailOnError("update revocation list", err)
	} else {
		//truncate password if requested
		if length > 0 && len(passwordStr) > length {
			passwordStr = passwordStr[0:length]
		}
		os.Stdout.Write([]byte(passwordStr))
		os.Stdout.Sync()
		//write the newline on stderr only, so that it is not copied when
		//piping stdout to xsel or xclip
		os.Stderr.Write([]byte("\n"))
	}
}

//FailOnError prints the given error and does not return (unless it is nil).
func FailOnError(operation string, err error) {
	if err != nil {
		os.Stderr.Write([]byte(operation + " failed: " + err.Error() + "\n"))
		os.Exit(1)
	}
}

//ParseArguments parses the os.Args. Will not return if they are malformed.
func ParseArguments() (domain string, length int, revoke bool) {
	usage := []byte("Usage: " + os.Args[0] + " [-r|--revoke] <domain> [length]\n")

	//read arguments
	var argRevoke = false
	var positionalArgs []string
	for _, arg := range os.Args[1:] {
		switch arg {
		case "-h", "--help":
			os.Stderr.Write(usage)
			os.Exit(0)
		case "-r", "--revoke":
			argRevoke = true
		default:
			positionalArgs = append(positionalArgs, arg)
		}
	}

	//need one or two arguments
	argDomain, argLength := "", 0
	switch len(positionalArgs) {
	case 1:
		argDomain = positionalArgs[0]
	case 2:
		argDomain = positionalArgs[0]
		var err error
		argLength, err = strconv.Atoi(positionalArgs[1])
		if err != nil {
			os.Stderr.Write([]byte(err.Error() + "\n"))
			os.Exit(1)
		}
	default:
		os.Stderr.Write(usage)
		os.Exit(1)
	}

	return argDomain, argLength, argRevoke
}

//GetMasterPassword queries the user for the master password.
func GetMasterPassword() ([]byte, error) {
	// read password from stdin if stdin is not a terminal
	if !terminal.IsTerminal(int(syscall.Stdin)) {
		return ioutil.ReadAll(os.Stdin)
	}

	//prompt is written to stderr because pwget may be used in a pipe where the
	//password is read from stdout by the next program (e.g. xsel or xclip)
	os.Stderr.Write([]byte("Master password: "))
	result, err := terminal.ReadPassword(int(syscall.Stdin))
	os.Stderr.Write([]byte("\n"))
	return result, err
}

func revocationListPath() string {
	return os.Getenv("HOME") + "/.pwget2-revocation"
}

//LoadRevocationList reads the revocation list, which contains the SHA256
//hashes of all passwords which have been revoked.
func LoadRevocationList() (map[string]bool, error) {
	//try to read file
	contents, err := ioutil.ReadFile(revocationListPath())
	if err != nil {
		if os.IsNotExist(err) {
			return map[string]bool{}, nil
		}
		return nil, err
	}

	//each hash is on one line
	hashes := strings.Split(string(contents), "\n")
	result := make(map[string]bool)
	for _, hash := range hashes {
		if hash != "" {
			result[hash] = true
		}
	}
	return result, nil
}

//AppendToRevocationList adds a new password hash to the revocation list.
func AppendToRevocationList(hash string) error {
	file, err := os.OpenFile(revocationListPath(), os.O_WRONLY|os.O_APPEND|os.O_CREATE, 0600)
	if err != nil {
		return err
	}

	_, err = file.Write([]byte(hash + "\n"))
	if err != nil {
		return err
	}

	return file.Close()
}

//Scrypt wraps scrypt.Key() and defines its parameters so that the KDF always
//produces the same results within the scope of this program.
func Scrypt(password, salt []byte) []byte {
	result, err := scrypt.Key(password, salt, 1<<16, 8, 16, 32)
	if err != nil {
		panic(err.Error())
	}
	return result
}

func z85EncodeToString(src []byte) string {
	dst := make([]byte, z85.EncodedLen(len(src)))
	n, err := z85.Encode(dst, src)
	if err != nil {
		panic(err.Error())
	}
	return string(dst[0:n])
}
